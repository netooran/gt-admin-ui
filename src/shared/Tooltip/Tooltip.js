import PropTypes from 'prop-types';
import { useEffect, useState } from 'react';
import styled from 'styled-components';
import AppConstants from '../../constants';

const Wrapper = styled.div`
  position: relative;
  display: inline-block;
  padding: 0.5rem 0;
  font-size: 0.75rem;

  & span {
    visibility: ${({ showTooltip }) => (showTooltip ? 'visible' : 'hidden')};
    width: 120px;
    background-color: black;
    color: #fff;
    text-align: center;
    padding: 0.5rem;
    border-radius: 6px;
    top: 100%;
    left: 50%;
    margin-left: -60px;

    position: absolute;
    z-index: 1;
  }

  &:hover span {
    visibility: visible;
  }

  & span::after {
    content: ' ';
    position: absolute;
    bottom: 100%;
    left: 50%;
    margin-left: -12px;
    border-width: 5px;
    border-style: solid;
    border-color: transparent transparent black transparent;
  }
`;

const Tooltip = ({ message, children, showTooltipOnLoad = false }) => {
  const [showTooltip, setShowTooltip] = useState(showTooltipOnLoad);

  useEffect(() => {
    showTooltip &&
      setTimeout(() => setShowTooltip(false), AppConstants.tootipActiveTime);
  }, [showTooltip]);

  return (
    <Wrapper showTooltip={showTooltip}>
      {children}
      <span>{message}</span>
    </Wrapper>
  );
};

Tooltip.propTypes = {
  message: PropTypes.string.isRequired,
  children: PropTypes.node.isRequired,
  showTooltipOnLoad: PropTypes.bool,
};

export default Tooltip;
