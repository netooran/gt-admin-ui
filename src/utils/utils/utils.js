export const noop = () => {};
export const delay = (ms) => new Promise((resolve) => setTimeout(resolve, ms));

const identity = (arg) => arg;

const capitalize = (str) => str.replace(/^\w/, (c) => c.toUpperCase());

const range = (limit) => [...Array(limit + 1).keys()].slice(1);

export const FunctionUtils = { identity };

export const StringUtils = { capitalize };

export const ArrayUtils = { range };
