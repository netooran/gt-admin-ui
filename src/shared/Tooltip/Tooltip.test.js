import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import AppConstants from '../../constants';
import { delay } from '../../utils/utils/utils';
import Tooltip from './Tooltip';

const tooltipText = 'You are hovering me';
const hoverText = 'Hover Me';

test('renders tooltip normally', async () => {
  const { container } = render(
    <Tooltip message={tooltipText}>{hoverText}</Tooltip>
  );

  expect(container).toMatchSnapshot();
  expect(screen.getByText(tooltipText)).not.toBeVisible();

  userEvent.hover(container);
  // FIX ME: Must assert on visibility of the tooltip on hover & unhover
  expect(screen.getByText(tooltipText)).toBeInTheDocument();
});

test(`renders hover content on load for ${AppConstants.tootipActiveTime}`, async () => {
  const { container } = render(
    <Tooltip message={tooltipText} showTooltipOnLoad>
      {hoverText}
    </Tooltip>
  );

  expect(container).toMatchSnapshot();
  expect(screen.getByText(tooltipText)).toBeVisible();

  await delay(AppConstants.tootipActiveTime + 1);
  expect(screen.getByText(tooltipText)).not.toBeVisible();
});
