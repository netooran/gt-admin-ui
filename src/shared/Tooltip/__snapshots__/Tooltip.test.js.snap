// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders hover content on load for 3000 1`] = `
.c0 {
  position: relative;
  display: inline-block;
  padding: 0.5rem 0;
  font-size: 0.75rem;
}

.c0 span {
  visibility: visible;
  width: 120px;
  background-color: black;
  color: #fff;
  text-align: center;
  padding: 0.5rem;
  border-radius: 6px;
  top: 100%;
  left: 50%;
  margin-left: -60px;
  position: absolute;
  z-index: 1;
}

.c0:hover span {
  visibility: visible;
}

.c0 span::after {
  content: ' ';
  position: absolute;
  bottom: 100%;
  left: 50%;
  margin-left: -12px;
  border-width: 5px;
  border-style: solid;
  border-color: transparent transparent black transparent;
}

<div>
  <div
    class="c0"
  >
    Hover Me
    <span>
      You are hovering me
    </span>
  </div>
</div>
`;

exports[`renders tooltip normally 1`] = `
.c0 {
  position: relative;
  display: inline-block;
  padding: 0.5rem 0;
  font-size: 0.75rem;
}

.c0 span {
  visibility: hidden;
  width: 120px;
  background-color: black;
  color: #fff;
  text-align: center;
  padding: 0.5rem;
  border-radius: 6px;
  top: 100%;
  left: 50%;
  margin-left: -60px;
  position: absolute;
  z-index: 1;
}

.c0:hover span {
  visibility: visible;
}

.c0 span::after {
  content: ' ';
  position: absolute;
  bottom: 100%;
  left: 50%;
  margin-left: -12px;
  border-width: 5px;
  border-style: solid;
  border-color: transparent transparent black transparent;
}

<div>
  <div
    class="c0"
  >
    Hover Me
    <span>
      You are hovering me
    </span>
  </div>
</div>
`;
